src/tools/kotlin-codegen-shared.ts

```diff
@@ -40,23 +41,30 @@ export function escapeIdentifier(name: string): string {
 /**
  * Generates a Kotlin type instance for the given handle connection.
  */
-export function generateConnectionType(connection: HandleConnection): string {
-  return generateConnectionSpecType(connection.spec, new SchemaGraph(connection.particle.spec).nodes);
+export function generateConnectionType(connection: HandleConnection,
+                                       schemaRegistry: {schema: Schema, generation: string}[] = []): string {
+  return generateConnectionSpecType(connection.spec, new SchemaGraph(connection.particle.spec).nodes, schemaRegistry);
 }
 
-export function generateConnectionSpecType(connection: HandleConnectionSpec, nodes: SchemaNode[]): string {
+export function generateConnectionSpecType(
+  connection: HandleConnectionSpec,
+  nodes: SchemaNode[],
+  schemaRegistry: {schema: Schema, generation: string}[] = []): string {
   let type = connection.type;
   if (type.isEntity || type.isReference) {
     // Moving to the new style types with explicit singleton.
     type = type.singletonOf();
   }
 
   return (function generateType(type: Type): string {
-    if (type.isEntity || type.isVariable) {
-      const node = type.isEntity
-        ? nodes.find(n => n.schema.equals(type.getEntitySchema()))
-        : nodes.find(n => n.variableName !== null && n.variableName.includes((type as TypeVariable).variable.name));
+    if (type.isEntity) {
+      const node = nodes.find(n => n.schema.equals(type.getEntitySchema()));
       return ktUtils.applyFun('EntityType', [`${node.fullName(connection)}.SCHEMA`]);
+    } else if (type.isVariable) {
+      const node = nodes.find(n => n.variableName !== null && n.variableName.includes((type as TypeVariable).variable.name));
```

Comment:
> `n.variableName.includes((type as TypeVariable).variable.name)`

.includes on a string does a substring check - I don't think you want that! Maybe you want simply:const node = `nodes.find(n => n.variableName === (type as TypeVariable).variable.name);` ?

Reply:
I think the thinking behind `include` was because of the nested variable names. In any case, I'll go for string equality.